analyzers/src/SonarAnalyzer.Common/Rules/ShouldImplementExportedInterfacesBase.cs

```diff
@@ -27,86 +27,109 @@
-        : ShouldImplementExportedInterfacesBase
-        where TArgumentSyntax : SyntaxNode
-        where TExpressionSyntax : SyntaxNode
-        where TClassSyntax : SyntaxNode
-    {
-        protected static bool IsOfExportType(ITypeSymbol type, INamedTypeSymbol exportedType) =>
+        protected abstract TSyntaxKind[] SyntaxKinds { get; }
+        protected abstract SeparatedSyntaxList<TArgumentSyntax>? GetAttributeArguments(TAttributeSyntax attributeSyntax);
+        protected abstract SyntaxNode GetAttributeName(TAttributeSyntax attributeSyntax);
+        protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+        protected abstract bool IsClassOrRecordSyntax(SyntaxNode syntaxNode);
+        protected abstract string GetIdentifier(TArgumentSyntax argumentSyntax);
+        protected abstract SyntaxNode GetExpression(TArgumentSyntax argumentSyntax);
+        // Retrieve the expression inside of the typeof()/GetType() (e.g. typeof(Foo) => Foo)
+        protected abstract SyntaxNode GetTypeOfOrGetTypeExpression(SyntaxNode expressionSyntax);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected ShouldImplementExportedInterfacesBase() =>
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, Language.RspecResources);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer,
+                c =>
+                {
+                    var attributeSyntax = (TAttributeSyntax)c.Node;
+
+                    if (!(c.SemanticModel.GetSymbolInfo(GetAttributeName(attributeSyntax)).Symbol is IMethodSymbol attributeCtorSymbol) || !attributeCtorSymbol.ContainingType.IsAny(exportAttributes))
+                    {
+                        return;
+                    }
+
+                    var exportedType = GetExportedTypeSymbol(GetAttributeArguments(attributeSyntax), c.SemanticModel);
+                    var attributeTargetType = GetAttributeTargetSymbol(attributeSyntax, c.SemanticModel);
+
+                    if (exportedType != null && attributeTargetType != null && !IsOfExportType(attributeTargetType, exportedType))
+                    {
+                        var action = exportedType.IsInterface()
+                            ? ActionForInterface
+                            : ActionForClass;
+
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, attributeSyntax.GetLocation(), action,
+                            exportedType.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat),
+                            attributeTargetType.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat)));
+                    }
+                },
+                SyntaxKinds);
```

Comment:
The SyntaxKinds property can be removed.
```suggestion
                Language.SyntaxKind.Attribute);
```